crypt.go

```diff
@@ -323,6 +454,100 @@ type ActivateVolumeOptions struct {
+	for _, err := range e.keyDataErrs {
+		fmt.Fprintf(&s, "\n- %v", err)
+	}
+	fmt.Fprintf(&s, "\nand activation with recovery key failed: %v", e.recoveryKeyUsageErr)
+	return s.String()
+}
+
+// ActivateVolumeWithKeyData attempts to activate the LUKS encrypted container at sourceDevicePath and create a
+// mapping with the name volumeName, using the supplied KeyData objects to recover the disk unlock key from the
+// platform's secure device. This makes use of systemd-cryptsetup.
+//
+// The ActivateOptions field of options can be used to specify additional options to pass to systemd-cryptsetup.
+//
+// If activation with the supplied KeyData objects fails, this function will attempt to activate it with the fallback
+// recovery key instead. The fallback recovery key will be requested using systemd-ask-password. The RecoveryKeyTries
+// field of options specifies how many attempts should be made to activate the volume with the recovery key before
+// failing. If this is set to 0, then no attempts will be made to activate the encrypted volume with the fallback
+// recovery key.
+//
+// If either the PassphraseTries or RecoveryKeyTries fields of options are less than zero, an error will be returned.
+// If the ActivateOptions field of options contains the "tries=" option, then an error will be returned. This option
+// cannot be used with this function.
+//
+// If activation with one of the supplied KeyData objects succeeds, a SnapModelChecker will be returned so that the
+// caller can check whether a particular Snap device model has previously been authorized to access the data on this
+// volume. If the fallback recovery key is used for activation, no SnapModelChecker will be returned.
+//
+// If activation fails, an error will be returned.
+func ActivateVolumeWithMultipleKeyData(volumeName, sourceDevicePath string, keys []*KeyData, options *ActivateVolumeOptions) (*SnapModelChecker, error) {
+	if len(keys) == 0 {
+		return nil, errors.New("no keys provided")
+	}
+	if options.PassphraseTries < 0 {
+		return nil, errors.New("invalid PassphraseTries")
+	}
+	if options.RecoveryKeyTries < 0 {
+		return nil, errors.New("invalid RecoveryKeyTries")
+	}
+
+	activateOptions, err := makeActivateOptions(options.ActivateOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	s := makeActivateWithKeyDataState(volumeName, sourceDevicePath, activateOptions, options.KeyringPrefix, keys)
+	switch s.run() {
+	case true: // success!
+		return s.snapModelChecker(), nil
+	default: // failed - try recovery key
+		if rErr := activateWithRecoveryKey(volumeName, sourceDevicePath, nil, options.RecoveryKeyTries, activateOptions, options.KeyringPrefix); rErr != nil {
+			// failed with recovery key - return errors
+			var kdErrs []error
+			for _, e := range s.errors() {
+				kdErrs = append(kdErrs, e)
+			}
+			return nil, &activateVolumeWithKeyDataError{kdErrs, rErr}
+		}
+		// succeeded with recovery key
+		return nil, nil
```

Comment:
the corresponding methods for TPM would list return error info about what failed with the protected keys even in this path, this new code doesn't, it's a bit confusing to have two approaches about this atm

Reply:
See https://github.com/snapcore/secboot/pull/134#discussion_r612570827 - I don't think it's useful to return an error in this case, and a future PR will add an alternative way to access this status.

Reply:
we have code right now that does that kind of tracking/reporting in snap-bootstrap, it doesn't care about exact errors information but it cares whether the recovery key was used  or not,  the relevant code in snapd is here:

https://github.com/snapcore/snapd/blob/master/secboot/secboot_tpm.go#L594

we need to be able to keep it working even switching to this in some form. Is used only for TPM atm but it would need to work also for this code when we have to wrap the hook code in this.  I don't think we can wait the new code you describe. So this is a regression from our POV.

Reply:
sorry, I was just confused, it seems  we can use *SnapModelChecker == nil to detect that the recover key was used, is that correct?

Reply:
I think my "confusion" relates to the fact as well that returning nil, nil like this is kind of atypical and fragile. It's very easy to write code that would panic on the nil pointer by assuming things that aren't true for this function but usually are.

Reply:
Yes, no SnapModelChecker is returned whenever the recovery key is used.

Reply:
@chrisccoulson basically I think even just a fixed ErrRecoverKeyUsed or similar would be more in line with common golang behavior patterns.

Reply:
I agree with Samuele here, to make our side less confusing I would be in favor of something similar to returning ActivateWithTPMSealedKeyError to keep the hook and TPM  parts somewhat similar (it's already hard to follow). Once we have this future alternative way to access the status we can then switch consistently. WDYT?

Reply:
I've added a ErrRecoveryKeyUsed error now.